lib/src/back_end/code_writer.dart

```diff
@@ -0,0 +1,198 @@
+    _options.indent = _pieceOptions[_pieceOptions.length - 2].indent + indent;
+  }
+
+  /// Increase the expression nesting of the current piece if [condition] is
+  /// `true`.
+  void nestIf(bool condition) {
+    if (!condition) return;
+
+    _options.nesting += Indent.expression;
+  }
+
+  /// Sets the number of spaces of expression nesting for code written by the
+  /// current piece to [nesting].
+  ///
+  /// Replaces any previous nesting set by this piece.
+  void setNesting(int nesting) {
+    // Include nesting from surrounding pieces.
+    _options.nesting =
+        _pieceOptions[_pieceOptions.length - 2].nesting + nesting;
+  }
+
+  /// Inserts a newline if [condition] is true.
+  ///
+  /// If [space] is `true` and [condition] is `false`, writes a space.
+  ///
+  /// If [indent] is given, sets the amount of block-level indentation for this
+  /// and all subsequent newlines to [indent].
+  void splitIf(bool condition, {bool space = true, int? indent}) {
+    if (indent != null) setIndent(indent);
+
+    if (condition) {
+      newline();
+    } else if (space) {
+      this.space();
+    }
+  }
+
+  /// Writes a single space to the output.
+  void space() {
+    write(' ');
+  }
+
+  /// Inserts a line split in the output. If [blank] is true, writes an extra
+  /// newline to produce a blank line.
+  void newline({bool blank = false}) {
+    handleNewline();
+    _finishLine();
+    _buffer.writeln();
+    if (blank) _buffer.writeln();
+
+    _column = _options.combinedIndentation;
+    _buffer.write(' ' * _column);
+  }
+
+  /// Sets whether newlines are allowed to occur from this point on for the
+  /// current piece of any of its children.
+  void setAllowNewlines(bool allowed) {
+    _options.allowNewlines = allowed;
+  }
+
+  /// Format [piece] and insert the result into the code.
+  void format(Piece piece) {
+    // Don't bother recursing into the piece tree if we know the solution will
+    // be discarded.
+    if (_containsInvalidNewline) return;
+
+    // TODO(tall): Sometimes, we'll want to reset the expression nesting for
+    // an inner piece, for when a block-like construct appears inside an
+    // expression. If it turns out that we don't actually need to handle indent
+    // and nesting separately here, then merge them into a single field.
+    _pieceOptions.add(_PieceOptions(
+        _options.indent, _options.nesting, _options.allowNewlines));
+
+    var state = _pieceStates.pieceState(piece);
+
+    // TODO(tall): Support pieces with different split costs, and possibly
+    // different costs for each state value.
+    if (state != 0) _cost++;
+
+    // TODO(perf): Memoize this. Might want to create a nested PieceWriter
+    // instead of passing in `this` so we can better control what state needs
+    // to be used as the key in the memoization table.
+    piece.format(this, state);
+
+    _pieceOptions.removeLast();
+  }
+
+  /// Format [piece] if not null.
```

Comment:
We can, yeah. It's just a convenience method.

As more and more of the language is supported, most of the of the formatter code will be:

* "Front-end" code in AstNodeVisitor and PieceFactory that traverses the AST and builds Pieces. That code mostly calls into the PieceWriter API. There will be a lot of this, so convenience methods in PieceWriter can be worth it.
* "Back-end" code in the `format()` method in Piece classes that traverses child pieces, inserts whitespace between them and writes code. That code mostly calls into CodeWriter here. There will be a lot of it, so convenience methods in CodeWriter can be worth it.

Think of PieceWriter is sort of the internal DSL for the AST->Piece translation rules, and CodeWriter as the internal DSL for the Piece->Text translation rules.

The surface area of both of those sets of rules is quite large (because Dart is a big language), so a lot of what I've been doing with the new IR is trying to figure out a nice API for the two DSLs so that it's easy and maintainable to pump out new Piece factory and format logic.